File: hw10/w9_wordplayer.py
# AUTHOR Shen_Han shawnhan@bu.edu
# AUTHOR Changlong_Jiang cljiang@bu.edu
import sys


def w9_wordplayer():
    ws = open(sys.argv[1]).read().split()
    bD = {}
    for k in range(len(ws)):
        tmp = list(ws[k])
        tmp.sort()
        mK = str(tmp)
        bD.setdefault(len(ws[k]), {mK: []}).setdefault(mK, []).append(ws[k])
    while(True):
        cmd = input().split()
        n = int(cmd[1])
        if n == 0:
            return 0
        probab = list(cmd[0])
        probab.sort()
        result = []
        if len(cmd[0]) == n:
            result = bD.get(n, {}).get(str(probab), [])
        if len(cmd[0]) > n:
            cur = list(range(0, n))
            flag = 0
            while(flag == 0):
                curMat = [' '] * n
                for jj in range(n):
                    curMat[jj] = probab[cur[jj]]
                curMat = str(curMat)
                result = (bD.get(n, {}).get(curMat, [])) + result
                for i in range(n - 1, -1, -1):
                    stop = len(probab) - n + i
                    tmp = cur[i]
                    cur[i] = stop + 1
                    while(tmp < stop):
                        tmp = tmp + 1
                        if probab[tmp] != probab[tmp - 1]:
                            cur[i] = tmp
                            break
                    if cur[i] > stop:
                        if i == 0:
                            flag = 1
                            break
                        continue
                    for kk in range(i + 1, n):
                        cur[kk] = cur[kk - 1] + 1
                    break
        result.sort()
        for c in result:
            print(c)
        print('.')

w9_wordplayer()

File: hw10/w9_wordplayer/w9_wordplayer/w9_wordplayer.cpp
#include<iostream>
#include<map>
#include<fstream>
#include<string>
#include<algorithm>
#include<vector>
using namespace std;

int mf(int a, int b) {
	int cc = 0;
	return 0;
}

int main(int argc, char const *argv[]) {
	ifstream wordlistFile;
	wordlistFile.open(argv[1]);
	string word;
	map<int, map<string, vector<string>>> bigDict;
	while (wordlistFile >> word) {
		string tmp = word;
		sort(tmp.begin(), tmp.end());
		bigDict[tmp.length()][tmp].push_back(word);
	}
	string check;
	int n;
	vector<string> result;
	while (1) {
		cin >> check >> n;
		int cc = mf(1, 2);
		if (n == 0)
			return 0;
		sort(check.begin(), check.end());
		//if (check.length() == n)
			//result = bigDict.find[n][check];
		if (check.length() > n) {
			int *cur = new int[n];
			for (int i = 0; i<n; i++)
				cur[i] = i;
			bool flag = true;
			while (flag) {
				string curMat;
				for (int i = 0; i < n; i++)
					curMat.push_back(check[cur[i]]);
				vector<string> tmp = bigDict[n][curMat];
				result.insert(result.end(), tmp.begin(), tmp.end());
				for (int i = n - 1; i > -1; i--) {
					int stop = check.length() - n + i;
					int tmp = cur[i];
					cur[i] = stop + 1;
					while (tmp < stop) {
						tmp++;
						if (check[tmp] != check[tmp - 1]) {
							cur[i] = tmp;
							break;
						}
					}
					if (cur[i] > stop) {
						if (i == 0) {
							flag = false;
							break;
						}
						continue;
					}
					for (int kk = i + 1; kk < n; kk++)
						cur[kk] = cur[kk - 1] + 1;
					break;
				}
			}
		}
		sort(result.begin(), result.end());
		for (int i = 0; i<result.size(); i++)
			cout << result[i] << endl;
		cout << '.' << endl;
	}
	return 0;
}
